src/cloud-api-adaptor/libvirt/config_libvirt.sh

```diff
@@ -130,3 +138,8 @@ case $TEST_E2E_SECURE_COMMS in
     echo "SECURE_COMMS=\"false\"" >> libvirt.properties
     ;;
 esac
+
+if [ ${OS_DISTRO} == "ubuntu" ]; then
+    # Reload shell so that pipx install PATH is available
+    exec $SHELL
```

Comment:
i'm curious. why don't we need this at line 87? and if we do, why don't we source the env directly after installation of pipx?

Reply:
Good question - I wrote this code in August, so can't complete remember the reason. I think it was because `pipx ensurepath` meant the path was only updated in the script's shell, which works for inside this script, but meant that the follow steps didn't have an updated path with the `exec bash`. Let me play with this again and try and refresh my memory to confirm.

I can just put this exec into line 87 instead of the ensurepath and see is that works as well?

Reply:
maybe I'm missing something, but the code looks like a noop. it re-sources .profile or something and then it's just discarded because the script (and the sh process, I assume).

Reply:
So I think I'm remembering the issues I saw that lead me to this:
When we run `pipx ensurepath` the output is:
```
/root/.local/bin has been been added to PATH, but you need to open a new terminal or re-login for this PATH change
    to take effect.

You will need to open a new terminal or re-login for the PATH changes to take effect.

Otherwise pipx is ready to go! ✨ 🌟 ✨
```
If we check our `PATH` then:
```
# echo $PATH
/usr/local/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```
It isn't added, but if we reload the shell and re-do it then it works:
```
# exec $SHELL
# echo $PATH
/usr/local/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/root/.local/bin
```

However re-loading the shell within the script means that all further output doesn't show up, so that's why I added the reload to the end of the script to avoid that loss of output.

I could remove it entirely and we'd then need to ask the user to reload their shell, so I chose to do it for them to remove a step.

Reply:
I don't fully understand this, so I've tried to give a cut down version of the script to demo:
```
# cat test.sh
pipx ensurepath
echo "PATH before reload: $PATH"
exec $SHELL
echo "We don't see this"
```
```
# ./test.sh
/root/.local/bin has been been added to PATH, but you need to open a new terminal or re-login for this PATH change
    to take effect.

You will need to open a new terminal or re-login for the PATH changes to take effect.

Otherwise pipx is ready to go! ✨ 🌟 ✨
PATH before reload: /usr/local/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
#
# echo $PATH
/usr/local/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/root/.local/bin
```

Reply:
ah, yeah this kind of message is pretty common. it installs something and adds it to the path in a .profile (or .bashrc or ...), but you don't have it in your path in the current shell, so the tool tells you to open a new shell or run their source script for the current shell.

> I could remove it entirely and we'd then need to ask the user to reload their shell, so I chose to do it for them to remove a step.

the user who is calling the script has to re-source .profile anyway, because the setting of the PATH is only scoped to the process that is 
invoking the script.

```
$ cat <<EOF | sh
#!/bin/sh
export PATH="bla:$PATH"
EOF
$ echo $PATH | grep bla

````

Reply:
yeah, but in a new shell, right? the new shell will load .profile (& friends) automatically

Reply:
Sorry, I still don't follow - when I leave the scope of the script why would it (without running `exec $SHELL` in that script) have a reload as I'm just returning to the shell scope I was already in before calling it?

Reply:
my point would be:

you can run `exec $SHELL` in the script, but that will only be valid in the child process that evaluates the script, not the script-invoking shell.

Reply:
So my point is that (as I've explained and showed example commands above), I would expect that behaviour, but it's not the case, hence the `exec $SHELL` is not a no-op

Reply:
Yeah this behaviour is for the local install, not the GHA, which is more straightforward as there is clear shell separation!